File: app/src/main/java/com/qweex/quite/QuiteAdapter.java
package com.qweex.quite;

import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.provider.MediaStore;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.util.Log;

import java.io.File;
import java.io.FilenameFilter;
import java.util.Arrays;

public class QuiteAdapter extends FragmentPagerAdapter {
    File dir;
    File[] files;
    int indexOfStart = 0;
    String startingFile;

    FragmentBase[] fragments;

    public QuiteAdapter(Uri target, FragmentManager fm, Context c) {
        super(fm);
        dir = target.getScheme().startsWith("content")
                ? new File(getRealPathFromUri(c, target))
                : new File(target.getPath());
        Log.d("dir", dir.getAbsolutePath() + "+" + dir.isFile() + "+" + dir.isDirectory());
        if(dir.isFile()) {
            startingFile = dir.getName();
            dir = dir.getParentFile();
            Log.d("Starting File", startingFile + " : " + dir.toString());
        }
        Log.d("Dir", dir.toString());
        files = dir.listFiles(new FilenameFilter() {
            @Override
            public boolean accept(File dir, String name) {
                return handlesFile(name, ImageFragment.filesHandled) ||
                        handlesFile(name, VideoFragment.filesHandled) ||
                        handlesFile(name, GifFragment.filesHandled);
            }
        });
        if(files==null) {
            files = new File[]{};
        }
        sort();
        fragments = new FragmentBase[files.length];
        try {
            for (int position = 0; position < files.length; position++) {
                String fullPath = files[position].getPath(); // = dir + "/" + files[position].getName();
                if (handlesFile(files[position].getPath(), ImageFragment.filesHandled))
                    fragments[position] = FragmentBase.newInstance(ImageFragment.class, fullPath);
                if (handlesFile(files[position].getPath(), VideoFragment.filesHandled))
                    fragments[position] = FragmentBase.newInstance(VideoFragment.class, fullPath);
                if (handlesFile(files[position].getPath(), GifFragment.filesHandled))
                    fragments[position] = FragmentBase.newInstance(GifFragment.class, fullPath);
            }
        }
        catch(Exception e) {}
    }


    public void sort() {
        if(Options.sortOrder==Options.RANDOM) {
        //    Collections.shuffle(files);
        } else
        Arrays.sort(files, Options.sortOrder);
        for(int position = 0; position < files.length; position ++)
            if (files[position].equals(startingFile))
                indexOfStart = position;
    }

    @Override
    public int getCount() {
        if(files==null)
            return 0;
        return files.length;
    }

    @Override
    public Fragment getItem(int position) {
        Log.d("Get Item", position + "!");
        return fragments[position];
    }

    //TODO: There's GOT to be a way to write this better
    boolean handlesFile(String filename, String[] handled) {
        for(String h : handled) {
            if(filename.toLowerCase().endsWith(h))
                return true;
        }
        return false;
    }

    public int getIndexOfStart() { return indexOfStart; }


    private static String getRealPathFromUri(Context context, Uri contentUri) {
        Cursor cursor = null;
        try {
            String[] proj = { MediaStore.Images.Media.DATA };
            cursor = context.getContentResolver().query(contentUri, proj, null, null, null);
            Log.d("WAAAA", contentUri + " " + proj + " " + cursor);
            int column_index = cursor.getColumnIndexOrThrow(MediaStore.Images.Media.DATA);
            cursor.moveToFirst();
            return cursor.getString(column_index);
        } finally {
            if (cursor != null) {
                cursor.close();
            }
        }
    }
}

File: app/src/main/java/com/qweex/quite/QuiteAdapter.kt
package com.qweex.quite

import android.content.Context
import android.database.Cursor
import android.net.Uri
import android.provider.MediaStore
import android.support.v4.app.FragmentManager
import android.support.v4.app.FragmentPagerAdapter
import android.support.v4.view.PagerAdapter
import android.util.Log
import java.io.File
import java.util.*

class QuiteAdapter(target: Uri, internal var fragmentManager: FragmentManager, c: Context) : FragmentPagerAdapter(fragmentManager) {
    internal var dir: File
    internal var files: ArrayList<File> = arrayListOf()
    var indexOfStart = 0
        internal set
    internal lateinit var startingFile: String

    internal lateinit var fragments: Array<FragmentBase?>

    init {
        dir = if (target.scheme!!.startsWith("content"))
            File(getRealPathFromUri(c, target))
        else
            File(target.path!!)
        Log.d("dir", dir.absolutePath + "+" + dir.isFile + "+" + dir.isDirectory)
        if (dir.isFile) {
            startingFile = dir.name
            dir = dir.parentFile
            Log.d("Starting File", startingFile + " : " + dir.toString())
        }
        Log.d("Dir", dir.toString())
        rescan()
    }

    private fun scanFolder(directory : File) {
        val foldersToScan : ArrayList<File> = arrayListOf()
        Log.d("rescan", "scan folder: " + directory.toString() + "...")
        val filesListed : Array<File> = directory.listFiles { subdir, name ->
            Log.d("rescan", "scanning file: " + name + "=" + handlesFile(name, Options.filetypesSelected))
            if(File(subdir,name).isDirectory and Options.recurse)
                foldersToScan.add(File(subdir, name))
            handlesFile(name, Options.filetypesSelected)
        }
        for(file in filesListed)
            files.add(file)
        for(fldr in foldersToScan)
            scanFolder(fldr)
    }

    fun rescan() {
        files.clear()
        scanFolder(dir)

        for (s in Options.filetypesSelected)
            Log.d("rescan", "filetype: " + s)

        if (Options.sortOrder === Options.RANDOM) {
            //    Collections.shuffle(files);
        } else {
            Collections.sort(files, Options.sortOrder)
            Log.d("rescan", "sort order: " + Options.sortOrder)
        }
        Log.d("rescan", "files count: " + files.size)
        for (f in files)
            Log.d("rescan", "file: " + f!!.name)
        fragments = arrayOfNulls(size = this.files.size)
        try {
            for (position in files.indices) {
                val fullPath = files[position].path // = dir + "/" + files[position].getName();
                if (handlesFile(files[position].path, ImageFragment.filesHandled))
                    fragments[position] = FragmentBase.newInstance(ImageFragment::class.java, fullPath)
                if (handlesFile(files[position].path, VideoFragment.filesHandled))
                    fragments[position] = FragmentBase.newInstance(VideoFragment::class.java, fullPath)
                if (handlesFile(files[position].path, GifFragment.filesHandled))
                    fragments[position] = FragmentBase.newInstance(GifFragment::class.java, fullPath)

                if (files[position].name == startingFile)
                    indexOfStart = position
            }
        } catch (e: Exception) {
        }
        this.notifyDataSetChanged()

    }

    override fun getCount(): Int {
        return files.size
    }

    override fun getItem(position: Int): FragmentBase? {
        Log.d("Get Item", position.toString() + "!")
        return fragments[position]
    }

    //TODO: There's GOT to be a way to write this better
    internal fun handlesFile(filename: String, handled: Array<String?>): Boolean {
        for (h in handled) {
            if (filename.toLowerCase().endsWith(h!!))
                return true
        }
        return false
    }


    private fun getRealPathFromUri(context: Context, contentUri: Uri): String {
        var cursor: Cursor? = null
        try {
            val proj = arrayOf(MediaStore.Images.Media.DATA)
            cursor = context.contentResolver.query(contentUri, proj, null, null, null)
            val column_index = cursor!!.getColumnIndexOrThrow(MediaStore.Images.Media.DATA)
            cursor.moveToFirst()
            return cursor.getString(column_index)
        } finally {
            cursor?.close()
        }
    }


    override fun getItemPosition(`object`: Any): Int {
        return PagerAdapter.POSITION_NONE
    }
}
